File: app/src/main/java/io/github/ovso/blackbox/data/network/BaseRequest.java
package io.github.ovso.blackbox.data.network;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public abstract class BaseRequest<T> {
  public final static int TIMEOUT_SECONDS = 7;
  public T getApi() {
    return createRetrofit().create(getApiClass());
  }

  protected abstract Class<T> getApiClass();

  protected abstract Headers createHeaders();

  protected abstract String getBaseUrl();

  private Retrofit createRetrofit() {
    Retrofit retrofit = new Retrofit.Builder().baseUrl(getBaseUrl())
        .addConverterFactory(GsonConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .client(createClient())
        .build();

    return retrofit;
  }

  private OkHttpClient createClient() {
    OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
    httpClient.readTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS);
    httpClient.connectTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS);
    httpClient.addInterceptor(new Interceptor() {
      @Override
      public Response intercept(Chain chain) throws IOException {
        Request original = chain.request();
        Request.Builder requestBuilder =
            original.newBuilder()
                .header("Content-Type", "plain/text")
                .headers(BaseRequest.this.createHeaders());
        Request request = requestBuilder.build();
        return chain.proceed(request);
      }
    });
    HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
    interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

    httpClient.addInterceptor(interceptor);
    OkHttpClient client = httpClient.build();
    return client;
  }

}

File: app/src/main/java/io/github/ovso/blackbox/data/network/BaseRequest.kt
package io.github.ovso.blackbox.data.network

import hu.akarnokd.rxjava3.retrofit.RxJava3CallAdapterFactory
import okhttp3.Headers
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.Retrofit.Builder
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

abstract class BaseRequest<T> {
  val api: T
    get() = createRetrofit().create(apiClass)

  protected abstract val apiClass: Class<T>

  protected abstract val baseUrl: String

  protected abstract fun createHeaders(): Headers

  private fun createRetrofit(): Retrofit {

    return Builder().baseUrl(baseUrl)
      .addConverterFactory(GsonConverterFactory.create())
      .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
      .client(createClient())
      .build()
  }

  private fun createClient(): OkHttpClient {
    val httpClient = OkHttpClient.Builder()
    httpClient.readTimeout(TIMEOUT_SECONDS.toLong(), TimeUnit.SECONDS)
    httpClient.connectTimeout(TIMEOUT_SECONDS.toLong(), TimeUnit.SECONDS)
    httpClient.addInterceptor { chain ->
      val original = chain.request()
      val requestBuilder = original.newBuilder()
        .header("Content-Type", "plain/text")
        .headers(this@BaseRequest.createHeaders())
      val request = requestBuilder.build()
      chain.proceed(request)
    }
    val interceptor = HttpLoggingInterceptor()
    interceptor.level = HttpLoggingInterceptor.Level.BODY

    httpClient.addInterceptor(interceptor)
    return httpClient.build()
  }

  companion object {
    const val TIMEOUT_SECONDS = 1000 * 60
  }

}
